src/init.c

```diff
@@ -832,6 +832,180 @@ void _julia_init(JL_IMAGE_SEARCH rel)
         jl_install_sigint_handler();
 }
 
+void jl_init_types2(void) JL_GC_DISABLED
+{
+    jl_module_t *core = NULL; // will need to be assigned later
+
+    jl_char_type = jl_new_primitivetype((jl_value_t*)jl_symbol("Char"), core,
+                                         jl_any_type, jl_emptysvec, 32);
+    jl_int8_type = jl_new_primitivetype((jl_value_t*)jl_symbol("Int8"), core,
+                                         jl_any_type, jl_emptysvec, 8);
+    jl_int16_type = jl_new_primitivetype((jl_value_t*)jl_symbol("Int16"), core,
+                                         jl_any_type, jl_emptysvec, 16);
+    jl_uint16_type = jl_new_primitivetype((jl_value_t*)jl_symbol("UInt16"), core,
+                                         jl_any_type, jl_emptysvec, 16);
+    jl_float16_type = jl_new_primitivetype((jl_value_t*)jl_symbol("Float16"), core,
+                                         jl_any_type, jl_emptysvec, 16);
+    jl_float32_type = jl_new_primitivetype((jl_value_t*)jl_symbol("Float32"), core,
+                                         jl_any_type, jl_emptysvec, 32);
+    jl_float64_type = jl_new_primitivetype((jl_value_t*)jl_symbol("Float64"), core,
+                                         jl_any_type, jl_emptysvec, 64);
+    jl_errorexception_type = jl_new_datatype(jl_symbol("ErrorException"), core, jl_type_type, jl_emptysvec,
+                                             jl_perm_symsvec(1, "msg"), jl_svec(1, jl_string_type), 0, 0, 1);
+}
+
+// Basic initialization that doesn't load a system image
```

Comment:
There seems to be some duplicate logic going on here. Could you generalize `_julia_init` to take the sysimage via a resource interface rather than expecting to find it in a file? Then various resource loaders could then be plugged in, for example:
* In the usual case, loading from file via some search paths
* An embedded binary blob like your mini sysimage
* Loading it over the network (maybe this could have benefits for wasm, or maybe it's just crazy talk.)

It might allow you to avoid duplicating the init logic so much?

Reply:
There is some duplicate logic, but I tried to minimize that by using the code in 'dump.c'. Your interface idea is interesting. I don't see how it minimizes duplication, though.

Reply:
Well I noticed that `jl_init_basics` shares some 100 lines of code with `_julia_init`. It seems like this could possibly be factored back together with some more flags or factored apart by extracting some of the shared code.

Reply:
Yeah, "factored apart" might be best. I'm worried about too many `if (standalone_aot_mode)` statements.

I'm still struggling with the resource interface. I understand what you mean, but I'm not sure how to code it, yet.